### PR TITLE
TileSet Python export plugins (fixes #1865)

### LINKED ISSUE #1865: TileSet Python export plugins
Hey Bjorn,

I'm currently using Tiled as my 'tileset' manager for Godot, where I make the maps in Godot and not Tiled, and it's going great. After messing around getting a good workflow between the two, I got pretty excited when I saw that there was Python scripting for export plugins, but after searching around I realised these weren't happening for tilesets, only for tilemaps. 

Is this on the roadmap as per #949?
Is this actually already implemented/possible but undocumented?
Do other things need to happen for this first, such as Python 2 -> 3, and if so, how can I help with this?

Kind regards,

### PR DESCRIPTION
**Pull request by DrDub**
Extended the Python plugin to support custom tileset exporters and readers written in Python.

I understand the Python plugin has its issues and the JavaScript plugin is preferred, but this extension might make it more useful. (The issue of system Python vs. Tiled Python sadly remains.)

### CODE REVIEW

File: src/plugins/python/pythonbind.cpp [hunk 1]
```diff
@@ -470,7 +477,7 @@ _wrap_PyQPoint_setY(PyQPoint *self, PyObject *args, PyObject *kwargs)
 
 
 PyObject *
-_wrap_PyQPoint_x(PyQPoint *self, PyObject *PYBINDGEN_UNUSED(_args), PyObject *PYBINDGEN_UNUSED(_kwargs))
+_wrap_PyQPoint_x(PyQPoint *self)
```
Comment (bjorn) [inline src/plugins/python/pythonbind.cpp:480]: Where do these changes come from? This file is normally generated using pybindgen, by executing `python tiledbinding.py`. When I do this with the latest version of pybindgen 0.22.1, it does not make these changes.

File: src/plugins/python/pythonplugin.cpp [hunk 2]
```diff
@@ -19,18 +19,23 @@
...
 #include <QDir>
 #include <QDirIterator>
 
 PyMODINIT_FUNC PyInit_tiled(void);
 extern int _wrap_convert_py2c__Tiled__Map___star__(PyObject *obj, Tiled::Map * *address);
+extern int _wrap_convert_py2c__Tiled__SharedTileset___star__(PyObject *obj, Tiled::SharedTileset * *address);
 extern PyObject* _wrap_convert_c2py__Tiled__Map_const___star__(Tiled::Map const * *cvalue);
 extern PyObject* _wrap_convert_c2py__Tiled__LoggingInterface(Tiled::LoggingInterface *cvalue);
+extern PyObject* _wrap_convert_c2py__Tiled__Tileset_const(Tiled::Tileset const *cvalue);
+extern PyObject* _wrap_convert_c2py__Tiled__LoggingInterface(Tiled::LoggingInterface *cvalue);
```
Comment (bjorn) [inline src/plugins/python/pythonplugin.cpp:38]: This line is now duplicated.

File: src/plugins/python/pythonplugin.cpp [hunk 3]
```diff
@@ -65,10 +71,16 @@ PythonPlugin::~PythonPlugin()
 {
     for (const ScriptEntry &script : mScripts) {
         Py_DECREF(script.module);
-        Py_DECREF(script.mapFormat->pythonClass());
+        if(script.mapFormat) {
+          Py_DECREF(script.mapFormat->pythonClass());
+        }
+        if(script.tilesetFormat) {
+          Py_DECREF(script.tilesetFormat->pythonClass());
+        }
```
Comment (bjorn) [inline src/plugins/python/pythonplugin.cpp:79]: Please slightly adjust the coding style `if (` and with 4 spaces for indentation.

File: src/plugins/python/pythonplugin.cpp [hunk 4]
```diff
@@ -192,7 +227,7 @@ void PythonPlugin::reloadModules()
 /**
  * Finds the first Python class that extends tiled.Plugin
```
Comment (bjorn) [inline src/plugins/python/pythonplugin.cpp:228]: Comment needs adjusting.
```suggestion
 * Finds the first Python class that extends the given \a pluginClass.
```

File: src/plugins/python/pythonplugin.cpp [hunk 5]
```diff
@@ -238,33 +273,53 @@ bool PythonPlugin::loadOrReloadModule(ScriptEntry &script)
...
+      }
     } else {
-        script.mapFormat = new PythonMapFormat(name, pluginClass, this);
-        addObject(script.mapFormat);
+      PyObject *module = PyImport_ImportModule(name.constData());
+      Py_DECREF(script.module);
+      script.module = module;
+      pluginClass = findPluginSubclass(script.module, mPluginClass);
+      if (!pluginClass) {
+          PySys_WriteStderr("Extension of tiled.Plugin not defined in "
```
Comment (bjorn) [inline src/plugins/python/pythonplugin.cpp:292]: Maybe adjust this to:
```suggestion
          PySys_WriteStderr("No extension of tiled.Plugin or tiled.TilesetPlugin defined in "
```

**Review by bjorn — CHANGES_REQUESTED** [review]
Even if this plugin isn't very well maintained and I would discourage writing new plugins in Python, I still think this is a very welcome addition. Thanks!

I've left a number of comments which need to be addressed before this can be merged.

File: src/plugins/python/pythonbind.cpp (hunk 1 above)
Comment (DrDub) [reply src/plugins/python/pythonbind.cpp:480]: You are right, sorry about it.  (I used the pybindgen shipped by Debian that's is 0.20.) 

Re-running it with the latest version.

File: src/plugins/python/pythonplugin.cpp (hunk 2 above)
Comment (DrDub) [reply src/plugins/python/pythonplugin.cpp:38]: fixed

File: src/plugins/python/tiledbinding.py [hunk 6]
```diff
@@ -406,6 +406,15 @@ def _decorate(obj, *args, **kwargs):
     foreign_cpp_namespace='Python',
     custom_name='Plugin')
 
+"""
+ C++ class PythonTilesetScript is seen as Tiled.TilesetPlugin from
```
[the remark this answers is not in the document]
Comment (DrDub) [reply src/plugins/python/tiledbinding.py:410]: you're correct, it's lowercase.

File: src/plugins/python/pythonplugin.cpp [hunk 7]
```diff
@@ -427,4 +486,92 @@ void PythonMapFormat::setPythonClass(PyObject *class_)
...
+            mError = tr("Script returned false. Please check console.");
+        return ret;
+    }
+
+    handleError();
+    return false;
+}
+
+
+void PythonTilesetFormat::setPythonClass(PyObject *class_)
```
[the remark this answers is not in the document]
Comment (DrDub) [reply src/plugins/python/pythonplugin.cpp:557]: They are different. One gets capabilities 

```cpp
            mCapabilities |= Tiled::MapFormat::Write;
            mCapabilities |= Tiled::MapFormat::Read;
```

the other gets capabilities

```cpp
            mCapabilities |= Tiled::TilesetFormat::Write;
            mCapabilities |= Tiled::TilesetFormat::Read;
```

but a more elegant solution might be possible, yes.

File: src/plugins/python/pythonplugin.cpp [hunk 8]
```diff
@@ -238,33 +273,53 @@ bool PythonPlugin::loadOrReloadModule(ScriptEntry &script)
...
+          PySys_WriteStdout("---- Tileset plugin\n");
+          script.tilesetFormat = new PythonTilesetFormat(name, pluginClass, this);
+          addObject(script.tilesetFormat);
+      }
     } else {
-        script.mapFormat = new PythonMapFormat(name, pluginClass, this);
-        addObject(script.mapFormat);
+      PyObject *module = PyImport_ImportModule(name.constData());
+      Py_DECREF(script.module);
+      script.module = module;
```
[the remark this answers is not in the document]
Comment (DrDub) [reply src/plugins/python/pythonplugin.cpp:289]: But removing those lines causes a SEGFAULT. Now I remember why I added them.

**Comment by bjorn** [comment]
@DrDub I've pushed some changes, and also was running into a crash, likely the same you encountered. I've fixed it by adding a check, but it does leave the plugin in a broken state in which it is not possible to actually implement a tileset format. When using your example plugin, I'm getting:
```
-- Loading tileset
TypeError: issubclass() arg 1 must be a class
No extension of tiled.Plugin or tiled.TilesetPlugin defined in script: tileset
```

**Review by bjorn — APPROVED** [review]
I think it's working fine now. Can you give it one last try @DrDub before we merge this?

**Pull request merged**